dbms/src/Functions/FunctionsRegexp.h

```diff
@@ -0,0 +1,807 @@
+            return String(const_string.data, const_string.size);
+        else
+            return String(reinterpret_cast<const char *>(&(*chars)[offsetAt(idx)]), sizeAt(idx) - 1);
+    }
+
+    void getStringRef(size_t idx, StringRef & dst) const
+    {
+        if constexpr (is_const)
+        {
+            dst.data = const_string.data;
+            dst.size = const_string.size;
+        }
+        else
+        {
+            auto tmp = StringRef(reinterpret_cast<const char *>(&(*chars)[offsetAt(idx)]), sizeAt(idx) - 1);
+            dst.data = tmp.data;
+            dst.size = tmp.size;
+        }
+    }
+
+    constexpr static bool isConst() { return is_const; }
+
+private:
+    size_t offsetAt(size_t i) const { return i == 0 ? 0 : (*offsets)[i - 1]; }
+    size_t sizeAt(size_t i) const { return i == 0 ? (*offsets)[0] : ((*offsets)[i] - (*offsets)[i - 1]); }
+
+    StringRef const_string;
+
+    // for vector string
+    const Chars_t * chars;
+    const Offsets * offsets;
+};
+
+template <bool is_const, typename T>
+class ParamInt
+{
+public:
+    DISALLOW_COPY_AND_MOVE(ParamInt);
+
+    // raise error in compile-time when type is incorrect
+    using Container = typename ColumnVector<std::enable_if_t<check_int_type<T>(), T>>::Container;
+
+    explicit ParamInt(Int64 val)
+        : const_int_val(val)
+        , int_container(nullptr)
+    {
+        if constexpr (!is_const)
+            throw Exception("non-const parm should not call this constructor");
+    }
+
+    // For passing compilation
+    explicit ParamInt(const StringRef &)
+        : const_int_val(0)
+        , int_container(nullptr)
+    {
+        throw Exception("Shouldn't call this constructor");
+    }
+
+    explicit ParamInt(const void * int_container_)
+        : const_int_val(0)
+        , int_container(reinterpret_cast<const Container *>(int_container_))
+    {
+        if constexpr (is_const)
+            throw Exception("const parm should not call this constructor");
+    }
+
+    // For passing compilation
+    ParamInt(const void *, const void *)
+        : const_int_val(0)
+        , int_container(nullptr)
+    {
+        throw Exception("Shouldn't call this constructor");
+    }
+
+    Int64 getInt(size_t idx) const
+    {
+        if constexpr (is_const)
+            return const_int_val;
+        else
+            return static_cast<Int64>((*int_container)[idx]);
+    }
+
+    String getString(size_t) const { throw Exception("ParamInt not supports this function"); }
+    void getStringRef(size_t, StringRef &) const { throw Exception("ParamInt not supports this function"); }
+    constexpr static bool isConst() { return is_const; }
+
+private:
+    Int64 const_int_val;
+
+    // for vector int
+    const Container * int_container;
+};
+
+// Columns may be const, nullable or plain vector, we can conveniently handle
+// these different type columns with Param.
+template <typename ParamImplType, bool is_nullable>
+class Param
+{
```

Comment:
`is_nullable == (null_map == nullptr)` is always true. 
So how about adding some exceptions in constructor functions.

Reply:
> `is_nullable == (null_map == nullptr)` is always true. So how about adding some exceptions in constructor functions.

okk, I will ensure it